### PR TITLE
feat(component): startScopedViewTranstion func to support scoped view transition

### PR DESCRIPTION
**Pull request by CatsJuice**
AF-1293

### CODE REVIEW

**Comment by CatsJuice** [comment]
* **#8093** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8093?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @CatsJuice and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/component/src/utils/view-transition.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  cb: () => Promise<void> | void,
+  options?: { timeout?: number }
+) {
+  if (typeof document === 'undefined') return;
+
+  if (typeof document.startViewTransition === 'function') {
+    const scopes = Array.isArray(scope) ? scope : [scope];
+    const timeout = options?.timeout ?? 20000;
+
+    scopes.forEach(setScope);
```
Comment (pengx17) [inline packages/frontend/component/src/utils/view-transition.ts:20]: only the last scope is applied to body?

Comment (CatsJuice) [reply packages/frontend/component/src/utils/view-transition.ts:20]: applied as `data-vt-scope-${scope}`

File: packages/frontend/mobile/src/components/search-input/style.css.ts [hunk 2]
```diff
@@ -1,10 +1,18 @@
 import { cssVarV2 } from '@toeverything/theme/v2';
-import { style } from '@vanilla-extract/css';
+import { generateIdentifier, style } from '@vanilla-extract/css';
+
+export const searchVTName = generateIdentifier('mobile-search-input');
+export const searchVTScope = generateIdentifier('mobileSearch');
```
Comment (pengx17) [inline packages/frontend/mobile/src/components/search-input/style.css.ts]: keep casing consistent?

File: packages/frontend/component/src/utils/view-transition.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+export function startScopedViewTransition(
+  scope: string | string[],
+  cb: () => Promise<void> | void,
+  options?: { timeout?: number }
+) {
+  if (typeof document === 'undefined') return;
+
+  if (typeof document.startViewTransition === 'function') {
+    const scopes = Array.isArray(scope) ? scope : [scope];
+    const timeout = options?.timeout ?? 20000;
```
Comment (pengx17) [inline packages/frontend/component/src/utils/view-transition.ts]: default timeout 20s seems too long

File: packages/frontend/component/src/ui/modal/modal.tsx [hunk 4]
```diff
@@ -86,21 +87,19 @@ class ModalTransitionContainer extends HTMLElement {
...
-            child.classList.add('vt-active');
-          }
-        });
-        document.startViewTransition(() => {
-          nodes.forEach(child => {
-            // eslint-disable-next-line unicorn/prefer-dom-node-remove
-            super.removeChild(child);
-          });
+          // eslint-disable-next-line unicorn/prefer-dom-node-remove
+          super.removeChild(child);
```
Comment (pengx17) [inline packages/frontend/component/src/ui/modal/modal.tsx:99]: why not use child.remove as unicorn suggested here?

File: packages/frontend/component/src/ui/modal/styles.css.ts [hunk 5]
```diff
@@ -7,10 +7,14 @@ import {
   keyframes,
   style,
 } from '@vanilla-extract/css';
+
+import { vtScopeSelector } from '../../utils/view-transition';
 export const widthVar = createVar('widthVar');
 export const heightVar = createVar('heightVar');
 export const minHeightVar = createVar('minHeightVar');
 
+export const modalVTScope = generateIdentifier('modal');
```
Comment (pengx17) [inline packages/frontend/component/src/ui/modal/styles.css.ts:16]: multiple modal share the same scope. Is it an issue?

File: packages/frontend/mobile/src/components/search-input/style.css.ts [hunk 6]
```diff
@@ -1,10 +1,18 @@
...
 
 export const wrapper = style({
   position: 'relative',
   backgroundColor: cssVarV2('layer/background/primary'),
-  viewTransitionName: 'mobile-search-input',
+
+  selectors: {
+    [`[data-vt-scope-${searchVTScope}] &`]: {
```
Comment (EYHN) [inline packages/frontend/mobile/src/components/search-input/style.css.ts]: vtScopeSelector ?

**Review by pengx17 — APPROVED** [review]

**Review by EYHN — APPROVED** [review]

**Pull request merged**